app/src/main/java/com/example/practice/MainActivityViewModel.kt

```diff
@@ -35,7 +35,7 @@ class MainActivityViewModel : ViewModel() {
     fun fetchUserList(excludingUserId: String = ""){
 
         // Call Network Api to get user data
-        RetrofitAPI().fetchUsersList(excludingUserId, {
+        API.create().fetchUsersList(excludingUserId, {
 
             // Perform task in coroutine as operation it time consuming
             viewModelScope.launch(Dispatchers.IO) {
```

Comment:
1. The success callback if `fetchUsers` is first handled inside `viewModelScope` using IO dispatch. Then execution is moved to `Main` dispatch to update `_userDataState`. What is the reason for each choice of dispatch? 
i.e. What is achieved by dispatching to IO, and what is achieved by dispatching from IO to Main afterwards?
2. In contrary, failure callback is not dispatched. Why is that ok? And what thread will failure be executing on considering its not explicitly dispatched to any thread?

Reply:
@nets-dev
Q1] The success callback if fetchUsers is first handled inside viewModelScope using IO dispatch. Then execution is moved to Main dispatch to update _userDataState. What is the reason for each choice of dispatch?
    i.e. What is achieved by dispatching to IO, and what is achieved by dispatching from IO to Main afterwards?
    **Ans** :- 1. The purpose of using Dispacter.IO is that we are performing excluding operations that are time-consuming. So if I did not use Dispacter.IO then we might face an ANR issue due to blocking UI thread.
                2. By using Dispacther.IO we perform a long time taking operations smoothly on a different thread and once the operation gets complete dispatch it on the Main thread using Dispatcher.Main.

Q2] In contrary, failure callback is not dispatched. Why is that ok? And what thread will failure be executing on considering its not explicitly dispatched to any thread?
    **Ans** :- fallback callback is not dispatched because it is handled in the main thread so no need to dispatch separately on the main thread.

Reply:
@lnt-nishantdande About the second question, what guarantees that failure callback will be invoked on main thread?

Reply:
@nets-dev For question 2, as api call is made from run-blocking without using any dispatcher/CoroutineScope. So run-blocking will run async API requests and wait for a response. Once the result/error is received, pass it to the success/failure function which is in the main thread. Hence it is guaranteed that a failure callback will be invoked on the main thread.